src/coreclr/nativeaot/Runtime/CommonMacros.h

```diff
@@ -177,12 +177,61 @@ typedef uint8_t CODE_LOCATION;
 // Define an unmanaged function called from managed code that needs to execute in co-operative GC mode. (There
 // should be very few of these, most such functions will be simply p/invoked).
 //
-#define COOP_PINVOKE_HELPER(_rettype, _method, _args) EXTERN_C NATIVEAOT_API _rettype REDHAWK_CALLCONV _method _args
-#ifdef HOST_X86
+#if defined(HOST_X86) && defined(HOST_WINDOWS)
+
+// x86 is special case. It supports multiple calling conventions (fastcall, stdcall, cdecl)
+// and mangles the method names according to the calling convention (eg. @fastcall@4, _stdcall@4,
+// _cdecl).
+//
+// The managed code uses its own calling convention that is different from the native call
+// conventions. It's similar to fastcall but pushes the arguments to stack in reverse order.
+// Additionally, for the sake of simplicity we don't decorate the symbol names.
+//
+// In order to bridge the managed calling convention we use two tricks:
+// - The COOP_PINVOKE_HELPER macro reorders parameters for any method with 4 or more arguments.
+// - A linker comment is used to pass the "/alternatename:foo=@foo@4" switch to allow the
+//   symbols to be resolved to the fastcall decorated name.
+
+#define COOP_ARGHELPER_NAME(_1, _2, _3, _4, _5, _6, NAME, ...) NAME
+#define COOP_ARGHELPER_NAME_(tuple) COOP_ARGHELPER_NAME tuple
+
+#define COOP_ARGHELPER0() ()
```

Comment:
This is exactly same as `FCDECL*` from CoreCLR. Could you please use the same name for it? (No need to physically share the macro definitions, just use the same macro name and shape.)

Reply:
It serves the same purpose but it's not the same. I used the varargs preprocessor magic to avoid having the argument count in the macro name. The goal was to reduce the number of changes and mostly keep the `COOP_PINVOKE_HELPER` usage as-is. If you prefer `FCDECL*` for consistency I can do it but I don't think it's necessarily better.

Reply:
I would like to +1 for `COOP_` name staying, it is much clearer name.

Reply:
Yes, I prefer consistency. There are number of arbitrary differences between native AOT and regular CoreCLR that should not be there. My ultimate goal is to have regular CoreCLR to be layered on top of native AOT base.

Reply:
> I would like to +1 for the COOP_ name staying (in usage), it is a much clearer name.

If you would like to give things better names, it is fine to rename it in both regular CoreCLR and native AOT.

Reply:
I committed something. Please let me know what you think. I'll split it off into separate PR as soon as we agree on the general shape.

There's `FCIMPL`, `FCIMPLEND`, and `FCDECL`. Unlike CoreCLR the `FCIMPL` macro doesn't have the variation for number of arguments and it's handled by preprocessor varargs expansion. This roughly corresponds to `FCIMPL[0-5]` and the same trick can be implemented in the CoreCLR as well if we want to unify it later. It doesn't work for non-integer-sized arguments (eg. double) where CoreCLR has special variations of the macros but that's not a concern for NativeAOT since the few places where it would be useful are written in assembly and anticipated to move to managed code anyway.

Analogously I renamed `PREEMPT_PINVOKE_HELPER` to `QCDECL` and `QCIMPL`. Strictly speaking we don't need that and CoreCLR doesn't use macros for those. It helps with marking methods that are called from managed side (as opposed to assembly helpers or other unmanaged code) though.

Reply:
I would prefer for the FCall and QCall shapes to be exactly same as in CoreCLR. There is no reason for why it should be different.

> QCDECL and QCIMPL

These are regular PInvokes with no special calling convention or argument mangling. I do not think we need to wrap the whole signature in the macros. Strictly speaking, even the `QCALLTYPE` that QCalls are tagged with in CoreCLR is unnecessary.

Reply:
> even the QCALLTYPE that QCalls are tagged with in CoreCLR is unnecessary.

But it is useful as a marker that the method is a QCall entrypoint.

Reply:
I can change the QCalls to match. I mainly used it as a marker in the beginning when there were calling convention mismatches all over the place.

How would you feel about adding `FCIMPL`/`FCDECL` without the argument count to CoreCLR and using it instead of `FCIMPL[0-n]`/`FCDECL[0-n]`?

Reply:
I’m not going to die on this hill. I don’t necessarily care about the cosmetic side when using the macro. At this point we also tossed the idea that with clever redefinition of the existing macro we can keep most of its usages intact. It was to reduce duplication in the macro definition this way but it’s ultimately small benefit.

Reply:
How about the version I committed now?

Reply:
Yes, it looks good to me.

> Drop NATIVEAOT_API from QCalls

I think this macro can be deleted from everywhere.

Reply:
Thanks. I'll split off the calling convention parts into a separate PR for review.

Reply:
https://github.com/dotnet/runtime/pull/99430